src/main.rs

```diff
@@ -1,91 +1,186 @@
+                    application_command
+                        .handle_command(&ctx, command)
+                        .await
+                        .expect("failed to handle application command");
+                    return;
+                }
+            }
 
-            // Create responses and shoot them off back to the channel
-            if let Err(why) = command
+            command
                 .create_interaction_response(&ctx.http, |response| {
-                    response
-                        .kind(InteractionResponseType::ChannelMessageWithSource)
-                        .interaction_response_data(|message| message.content(content))
+                    response.interaction_response_data(|message| {
+                        message.content(format!(
+                            "Unhandled application command \"{}\"!",
+                            command.data.name
+                        ))
+                    })
                 })
                 .await
+                .expect("failed to send message");
+        }
+    }
+
+    // Dispatched when a guild is created; or an existing guild's data is sent to
+    // us.
+    // Provides the guild's data and whether the guild is new.
+    //
+    // When we detect that a guild was created (either an existing guild has just
+    // become available to this bot or a guild was actually created),
+    // register all [`APPLICATION_COMMANDS`] under the newly detected guild.
+    //
+    // # Panics
+    //
+    // Panics if there was an error registering any of [`APPLICATION_COMMANDS`].
+    async fn guild_create(&self, ctx: Context, guild: Guild, _is_new: bool) {
+        for application_command in APPLICATION_COMMANDS {
             {
-                println!("Cannot respond to slash command: {}", why);
+                application_command
+                    .register_command(&ctx, &guild)
+                    .await
+                    .expect("failed to register application command");
+                println!(
+                    "Successfully registered application command: {}",
+                    application_command.get_name()
+                );
             }
         }
     }
 }
 
+/// We expect APPLICATION_ID and DISCORD_TOKEN for the bot application to be
+/// passed to the program at startup, via environment variables.
 #[tokio::main]
-async fn main() {
-    // Load framework
-    let framework = StandardFramework::new()
-        .configure(|c| c.prefix("!"))
-        .group(&GENERAL_GROUP);
-
-    // The Application Id is usually the Bot User Id.
-    let application_id: u64 = fs::read_to_string("app_id")
-        .expect("Could not read file.")
+async fn main() -> Result<(), Box<dyn Error>> {
+    let application_id: u64 = env::var("APPLICATION_ID")
+        .map_err(|_| "Expected APPLICATION_ID in the environment")?
         .parse()
-        .expect("Something went wrong reading the file");
+        .map_err(|_| "APPLICATION_ID is not a valid id")?;
```

Comment:
what's the benefit of a map_err?

Reply:
I'm rather adverse to panicking in general. `map_err` here gives us a more descriptive error message, while still letting us utilize the "returning `result` from main" pattern.

If you find it more readable, I can just use `.expect` here instead.

Reply:
map_err is fine, I was just curious since I haven't seen this pattern before.